### PR TITLE
Convert Markdown to TypeScript

### PR DESCRIPTION
**Pull request by dbkr**
and generally fix lint errors

### CODE REVIEW

File: src/Markdown.ts [hunk 1]
```diff
@@ -15,16 +16,16 @@ limitations under the License.
 */
 
 import * as commonmark from 'commonmark';
```
Comment (t3chguy) [inline src/Markdown.ts:18]: If you install https://www.npmjs.com/package/@types/commonmark we'll get access to all its goodies

File: src/Markdown.ts [hunk 2]
```diff
@@ -15,16 +16,16 @@ limitations under the License.
...
-import {escape} from "lodash";
+import { escape } from "lodash";
 
 const ALLOWED_HTML_TAGS = ['sub', 'sup', 'del', 'u'];
 
 // These types of node are definitely text
 const TEXT_NODES = ['text', 'softbreak', 'linebreak', 'paragraph', 'document'];
 
-function is_allowed_html_tag(node) {
+function isAllowedHtmlTag(node: any) {
```
Comment (t3chguy) [inline src/Markdown.ts:26]: the typing lib will give us this Node type here, but also `: boolean` plz

File: src/Markdown.ts [hunk 3]
```diff
@@ -39,21 +40,12 @@ function is_allowed_html_tag(node) {
...
-    }
-}
-
 /*
  * Returns true if the parse output containing the node
  * comprises multiple block level elements (ie. lines),
  * or false if it is only a single line.
  */
-function is_multi_line(node) {
+function isMultiLine(node) {
```
Comment (t3chguy) [inline src/Markdown.ts:48]: types please

File: src/Markdown.ts [hunk 4]
```diff
@@ -65,8 +57,13 @@ function is_multi_line(node) {
  * Class that wraps commonmark, adding the ability to see whether
  * a given message actually uses any markdown syntax or whether
  * it's plain text.
+ *
+ * Types are a bit of a struggle here as commonmark doesn't have types
  */
 export default class Markdown {
+    private input: string;
+    private parsed: any;
```
Comment (t3chguy) [inline src/Markdown.ts:65]: hopefully one you can pluck from @types/commonmark

**Review by t3chguy — CHANGES_REQUESTED** [review]
Bunch of missing types which should be easily plucked from @types/commonmark which needs installing

File: src/Markdown.ts (hunk 1 above)
Comment (dbkr) [reply src/Markdown.ts:18]: Ah, it *is* a thing. We still have to expose the methods for the individual types of element, but it's better.

File: src/Markdown.ts [hunk 5]
```diff
@@ -150,19 +153,26 @@ export default class Markdown {
             }
         };
 
-        renderer.html_inline = html_if_tag_allowed;
+        renderer.html_inline = function(node: any) {
```
Comment (t3chguy) [inline src/Markdown.ts:156]:
```suggestion
        renderer.html_inline = function(node: commonmark.Node) {
```

File: src/Markdown.ts [hunk 6]
```diff
@@ -118,16 +121,16 @@ export default class Markdown {
         //
         // Let's try sending with <p/>s anyway for now, though.
 
-        const real_paragraph = renderer.paragraph;
+        const realParagraph = renderer.paragraph;
 
         renderer.paragraph = function(node, entering) {
```
Comment (t3chguy) [inline src/Markdown.ts:126]:
```suggestion
        renderer.paragraph = function(node: commonmark.Node, entering: boolean) {
```

total guess

File: src/Markdown.ts [hunk 7]
```diff
@@ -150,19 +153,26 @@ export default class Markdown {
...
+        renderer.html_inline = function(node: any) {
+            if (isAllowedHtmlTag(node)) {
+                this.lit(node.literal);
+                return;
+            } else {
+                this.lit(escape(node.literal));
+            }
+        };
 
         renderer.html_block = function(node) {
```
Comment (t3chguy) [inline src/Markdown.ts:165]:
```suggestion
        renderer.html_block = function(node: commonmark.Node) {
```

File: src/Markdown.ts [hunk 8]
```diff
@@ -177,14 +187,13 @@ export default class Markdown {
      * N.B. this does **NOT** render arbitrary MD to plain text - only MD
      * which has no formatting.  Otherwise it emits HTML(!).
      */
-    toPlaintext() {
-        const renderer = new commonmark.HtmlRenderer({safe: false});
-        const real_paragraph = renderer.paragraph;
+    toPlaintext(): string {
+        const renderer = new commonmark.HtmlRenderer({ safe: false }) as CommonmarkHtmlRendererInternal;
 
         renderer.paragraph = function(node, entering) {
```
Comment (t3chguy) [inline src/Markdown.ts:193]:
```suggestion
        renderer.paragraph = function(node: commonmark.Node, entering: boolean) {
```

total guess

File: src/Markdown.ts [hunk 9]
```diff
@@ -193,7 +202,7 @@ export default class Markdown {
 
         renderer.html_block = function(node) {
```
Comment (t3chguy) [inline src/Markdown.ts:203]:
```suggestion
        renderer.html_block = function(node: commonmark.Node) {
```

File: src/Markdown.ts (hunk 6 above)
Comment (dbkr) [reply src/Markdown.ts:126]: yeah, looks right

**Review by t3chguy — APPROVED** [review]
:shipit:

File: package.json [hunk 10]
```diff
@@ -54,6 +54,7 @@
   },
   "dependencies": {
     "@babel/runtime": "^7.12.5",
+    "@types/commonmark": "^0.27.4",
```
Comment (jryans) [inline package.json:57]: We may want to move this to dev deps to match the others.

**Pull request merged**